Add Promises & Signing to Wallet Classes

skubakdj opened the pull request:
### What This PR Does

Inspired by the discussion from issue #103 a few days ago:

* Implements promise-based methods on the `BaseWallet` & `PrivKeyWallet` classes
* Creates new `signing.js` file that handles private key TX & message signing
* Adds raw TX validation to `validators.js`
* Implements `signRawTransaction` and `signMessage` methods on `PrivKeyWallet` class
* Adds dependency to `ethereumjs-tx` in `package.json` (used in v3)

Transaction signing was modeled from v3 file [uiFuncs.js](https://github.com/kvhnuke/etherwallet/blob/3ab97999a157b1e0410de49b40080bcd35a1f5b3/app/scripts/uiFuncs.js) (~line 96).

Message signing was modeled from v3 file [signMsgCtrl.js](https://github.com/kvhnuke/etherwallet/blob/75157b1279f31d13ab975af97eefbed098265990/app/scripts/controllers/signMsgCtrl.js) (~line 22).

### What It Doesn't Do

It does not update our `package-lock.json` file. Just doing a `npm install` after pulling down a completely fresh copy of the repo caused a few packages to bump up in version numbers. I'd like to first understand why, though can commit the changes if necessary.

dternyak commented on common/libs/signing.js:
```diff
@@ -0,0 +1,39 @@
+    throw new Error('Invalid raw transaction');
+  }
+
+  let eTx = new EthTx(rawTx);
```
V3 Implementation for reference: https://github.com/kvhnuke/etherwallet/blob/3ab97999a157b1e0410de49b40080bcd35a1f5b3/app/scripts/uiFuncs.js#L113

dternyak commented on common/libs/validators.js:
```diff
@@ -84,3 +84,45 @@ export function isPositiveIntegerOrZero(number: number): boolean {
+
+    const value = rawTx[prop.name];
+
+    if (!rawTx.hasOwnProperty(prop.name)) return (valid = false);
```
Lots of unnecessary returns here -- any way this could be cleaned up?

skubakdj replied:
Yep, I'll figure something else out

wbobeirne replied:
Honestly this seems like a good use-case for the good ol' fashioned for loop, to spare dealing with the `valid` variable and just return out of the function at the first sign of badness.

skubakdj replied:
Whoops, pushed a fix before I read your comment. A for loop would indeed simplify things :+1:

wbobeirne commented on common/libs/signing.js:
```diff
@@ -0,0 +1,39 @@
+import { sha3, ecsign } from 'ethereumjs-util';
+import { isValidRawTx } from 'libs/validators';
+
+export function signRawTxWithPrivKey(privKey: Buffer, rawTx: Object): string {
```
We should define a rawTx flow type, especially since there's a function that outlines what one should have.

wbobeirne commented on common/libs/wallet/base.js:
```diff
@@ -1,11 +1,15 @@
 
 export default class BaseWallet {
-  getAddress(): string {
+  getAddress(): Promise<any> {
```
If it's going to get implemented as a promise elsewhere, this should probably return a promise that rejects, rather than throwing.

wbobeirne commented on common/libs/wallet/privkey.js:
```diff
@@ -31,7 +34,37 @@ export default class PrivKeyWallet extends BaseWallet {
+  }
+
+  unlock(): Promise<any> {
+    return new Promise(resolve => {
```
You can just return `Promise.resolve()`

skubakdj commented:
@wbobeirne addressed your feedback with this latest round of commits. Also went with the for loop over forEach, though I could go either way on this if anyone has a preference :smile:

wbobeirne commented:
Aces, thanks! Normally I wouldn't care either, but that's literally the only case where a regular for loop is more useful. I really wish forEach had a way to break out...
